sync/refcount.go

```diff
@@ -0,0 +1,59 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+)
+
+type ReleaseDataFunc = func(ctx context.Context, data interface{}) error
+
+type RefCounter struct {
+	count           int64
+	data            interface{}
+	releaseDataFunc ReleaseDataFunc
+}
+
+// Data returns data
+func (r *RefCounter) Data() (interface{}, error) {
+	v := atomic.LoadInt64(&r.count)
+	if v <= 0 || r.data == nil {
+		return nil, fmt.Errorf("using RefCounterer after data released")
```

Comment:
This error can be determined by static code analysis, thus should raise a panic instead.